src/scanpy/_singleton.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
+
+import os
+from traceback import extract_stack
+from types import FunctionType, MethodType
+
+
+def documenting() -> bool:
+    """Return whether this is being called from Sphinx.
+
+    (but not when e.g. a log message is accessing `settings.verbosity`)
+    """
+    if not os.environ.get("SPHINX_RUNNING"):
+        return False
+    found = False
+    logging = False
+    for frame in extract_stack():
+        # Let any sphinx ext get the docstring
+        if frame.name in {
+            "eval_config_file",  # Sphinx import
+            "generate_autosummary_docs",  # Autosummary generator
+            # "parse_generated_content",  # Autodoc parser
+            "get_object_members",  # Class level of autodoc
+            "import_object",  # Attr level of autodoc
+        }:
+            found = True
+        if frame.filename.endswith("/logging.py"):
+            logging = True
+    return found and not logging
+
+
+class SingletonMeta(type):
+    def __new__(mcls, cls_name: str, *args, **kwargs):
+        cls = super().__new__(mcls, cls_name, *args, **kwargs)
+
+        # We do something differently when we are imported by autosummary.
+        if documenting():
+            props = {}
+            for name in dir(cls):
+                if name.startswith("_") or name in cls.__dict__:
```

Comment:
Why isn't the second half of this check looking for some sort of deprecated marker? I only see `cls.set_figure_params` as the only member of `cls.__dict__` that doesn't ebigin with `_` but also that would be filtered out.  However, I wouldn't say that it should be filtered out because it is in `__dict__` but rather because it is deprecated.